### ISSUE TITLE
Bulk domain / user creation

### ISSUE THREAD

**Issue by llopat** [nmaas-platform/nmaas-platform#19]
The system administrator (or a user with a new specific role - not yet defined) should have the possibility to trigger a creation of several domains and assignment of users to those domains within a particular domain with a single call.
Specification of the domains and users to be deployed should be provided in a csv file (exact schema to be designed).

The platform should expose additional API endpoint to trigger the creation.
Note: the concept of a group of domain will be implemented under a dedicated ticket

Some initial requirements for this new process are as follows:
- a list of users is uploaded with a set of required fields (to be determined) including the domain group(s) in which they should join
- the missing domain groups are created (what would be some sane defaults here, e.g. for the initial list of allowed applications?)
- the missing users are created, along with a corresponding domain to which they are assigned as domain managers. Each domain is joined to the respective domain groups as specified in the CSV file.
- in case a user already exists with the same ID (e.g. username), the user creation and domain creation processes are skipped, but we still assign their existing domain to the specified domain groups (UNION with existing domain groups which they are already part of).

**Comment by llopat** [comment]
@korvoj please provide additional insight, you can update my initial comment if required

**Comment by korvoj** [comment]
As an initial set of accepted parameters in the CSV file, we can evaluate the use of:

- domain name [required] (is it possible for a domain name to only be consisted of numbers currently, e.g. `12345` ? This can be beneficial since some institutions use only numbers for their student IDs, and such an ID number is a good candidate for unique identifying the personal domain of a user) 
- username of the domain administrator [required] (would it be possible to only create a "dummy" record in case the user does not yet exist when the bulk creation of domains is attempted? The username of the user once registered is expected to be known before hand, since it will be provided by the IdP (e.g., firstname.lastname). This would allow the user to immediately have access to their previously provisioned domain once they log in for the first time.
- definition of DC networks [optional] (in the general case this would be the client-access VPN IP for the given user. This can be beneficial as to keep track of which IPs are assigned to which users and we might find use for this information in the future such as using it as a Helm parameter for creation of NetworkPolicies during application deployment).
- a list of domain groups [optional] (this would be a list of domain groups to which the domain should be added. The list of domain groups can use a single column in the CSV file by utilizing a different separator)

**Comment by llopat** [comment]
@korvoj thanks for the input, please find my further questions below:
- in general, what is the plan for actual domain / namespace creation inside the K8s cluster? We'd like to have this automated, right?
- does K8s allow for names of namespaces to container only numbers? For the namespace we can either re-use the domain codename or (if not the same) use additional dedicated parameter from the CSV.
- creation of a dummy user account sounds doable and reasonable. This would only apply to users that would later on log in with SS/IdP or "local" accounts as well?

**Comment by korvoj** [comment]
> in general, what is the plan for actual domain / namespace creation inside the K8s cluster? We'd like to have this automated, right?

Yes, we should discuss if an Ansible playbook can be invoked during the domain creation process. Alternatively, we can think if it would be possible to model the domain creation process using a Helm chart (e.g., deploying a dedicated Helm chart for new domain creation that would create the necessary ingress controller, set up the MetalLB configuration...).

> does K8s allow for names of namespaces to container only numbers? For the namespace we can either re-use the domain codename or (if not the same) use additional dedicated parameter from the CSV.

Yes, Kubernetes allows creation of namespaces whose names are only numbers. I would like to keep the set of fields in the CSV file to a minimum, as not to overburden the person doing the bulk creation. Let's try to keep the domain codename same as the Kubernetes namespace for now, and we can customize this later if such a requirement arises.

> creation of a dummy user account sounds doable and reasonable. This would only apply to users that would later on log in with SS/IdP or "local" accounts as well?

The scenario that I had in mind for this feature was when using an IdP. However, if the implementation is not too complicated to accommodate local accounts as well, we should add that flexibility. Integration with an IdP might be a more involved process in some cases, since it is usually managed by a different set of people than the ones who would deploy NMaaS.

**Comment by kbeyro** [comment]
@korvoj @llopat 
`definition of DC networks [optional] (in the general case this would be the client-access VPN IP for the given user. This can be beneficial as to keep track of which IPs are assigned to which users and we might find use for this information in the future such as using it as a Helm parameter for creation of NetworkPolicies during application deployment).` 

Do we have already that kind of fields in User/Domain view somewhere ? If not, this field should only have information about IP or would you like to have some more information ?

**Comment by llopat** [comment]
@kbeyro we had some field in the Domain view but recently it was hidden in the Portal since is wasn't used and might have been confusing for the users. So it should still be there in the data model. Maybe not a priority for the first version of the prototype.

**Comment by llopat** [comment]
@korvoj some questions and comments regarding current implementation
- in case of bulk domain creation would you like to see a two step process, like first uploading the csv and receiving an output from the platform summarizing the objects (domains, users, groups) that will be actually created and a second step for final confirmation and trigger of actual domain creation?
- do you foresee the need for "dynamic" column name mapping of the csv (as mentioned for bulk app deployments) or we can assume that all the required columns with known static names will be provided in the csv?
- do you see the need for tracking "domain/user bulks" as separate entities? Meaning that user from the Portal is able to see the status of given bulk deployment and e.g. trigger bulk removal at some point (I think this will be a must in case of bulk app deployments but not sure if same for domains/users)

**Comment by llopat** [comment]
@korvoj also what about the automated namespace creation? If you could prepare an Issue for this we could have someone to pick this up on our dev side (like triggering Ansible or whatever technology you chose on your side)

**Comment by korvoj** [comment]
@llopat ,

> in case of bulk domain creation would you like to see a two step process, like first uploading the csv and receiving an output from the platform summarizing the objects (domains, users, groups) that will be actually created and a second step for final confirmation and trigger of actual domain creation?

Yes, it will be very beneficial if it is implemented in this way. We might even think about extending it to a three step process, where the third step is actually a progress bar indicating the current status of the bulk domain creation/application deployment. This visual feedback might be very beneficial, taking into account that bulk deployment of 100s of apps or creating 100s of new domains might take quite some time. Preferably, it would be possible to navigate away and come back to this progress tracker (e.g., a dedicated page).

> do you foresee the need for "dynamic" column name mapping of the csv (as mentioned for bulk app deployments) or we can assume that all the required columns with known static names will be provided in the csv?
Yes, I think this can make the process of bulk deployment much easier from a user/admin perspective, eliminating the problem of typos and ambiguities related to spelling/capitalization, and column names. It will also ease the compatibility with different LMSs since, they all store the same information (names, emails, id numbers...) but might expose them differently in exported files (thus requiring users to manually change the columns before uploading them to NMaaS).

> do you see the need for tracking "domain/user bulks" as separate entities? Meaning that user from the Portal is able to see the status of given bulk deployment and e.g. trigger bulk removal at some point (I think this will be a must in case of bulk app deployments but not sure if same for domains/users)
Yes, bulk removal of deployments would be very beneficial, and I completely agree with you. When it comes to bulk removal of domains/users, that might also be beneficial (e.g., removing domains, once the student graduates and completes their studies), but it might not be such a high priority at the moment. 

Regarding the automated namespace creation, this can be done in a variety of ways, so I think it would be best to first sync during some VC, before coming up with the formal requirements.

**Comment by llopat** [comment]
@korvoj let's keep this issue open and move it to the scope of v1.7.0. Most of the features are already addressed but let's see what is still missing after we release v1.6.0
